ocaml/xapi/xapi_vm_snapshot.ml

```diff
@@ -348,13 +348,16 @@ let extended_do_not_copy = [
   "children";
 ] @ do_not_copy
 
+let domain_type_to_string = function
+  | `pv -> "pv"
+  | `hvm -> "hvm"
+
 (* Update the domain_type field if it does not match the HVM_boot_policy *)
-let ensure_domain_type_is_consistent ~__context ~vm =
-  Db.VM.get_HVM_boot_policy ~__context ~self:vm
-  |> fun p -> Xapi_vm_helpers.derive_domain_type ~hVM_boot_policy:p
-  |> fun value ->
-    if Db.VM.get_domain_type ~__context ~self:vm <> value then
-      Db.VM.set_domain_type ~__context ~self:vm ~value
+let ensure_domain_type_is_consistent ~__context ~snap_metadata =
+  match (List.assoc_opt "domain_type" snap_metadata) with
+  | Some `unspecified -> ("domain_type", domain_type_to_string (Xapi_vm_helpers.derive_domain_type (List.assoc "HVM_boot_policy" snap_metadata))) :: List.remove_assoc "domain_type" snap_metadata
```

Comment:
it'd be nice to optimize that - i assume `domain_type` often exists in `snap_metadata` and doesn't change too much

Reply:
it's a shame `snap_metadata` isn't a hashtable

Reply:
I think the common case will be that domain type is present and specified, the unspecified and None cases are the fallbacks.